File: web/src/api/watchlist.ts
import apiClient from "./client";
import type { WatchlistItem, WatchlistSummary } from "@/types/market";

export interface WatchlistResponse {
  items: WatchlistItem[];
  summary: WatchlistSummary;
}

export async function fetchWatchlist(): Promise<WatchlistResponse> {
  const { data } = await apiClient.get<WatchlistResponse>("/watchlist");
  return data;
}

export async function addWatchlistItem(
  payload: Pick<
    WatchlistItem,
    "code" | "name" | "price" | "cost" | "changePct" | "pnl" | "holdingValue" | "positionPct"
  >
): Promise<WatchlistItem> {
  const { data } = await apiClient.post<WatchlistItem>("/watchlist", payload);
  return data;
}

export async function updateWatchlistItem(
  code: string,
  patch: Partial<WatchlistItem>
): Promise<WatchlistItem> {
  const { data } = await apiClient.put<WatchlistItem>(
    `/watchlist/${code}`,
    patch
  );
  return data;
}

export async function deleteWatchlistItem(code: string): Promise<void> {
  await apiClient.delete(`/watchlist/${code}`);
}

File: api/app/models/watchlist.py
"""自选/持仓表"""

from typing import Optional
from sqlmodel import SQLModel, Field


class WatchlistItem(SQLModel, table=True):
    """自选池个股"""
    __tablename__ = "watchlist"

    id: Optional[int] = Field(default=None, primary_key=True)
    # 唯一约束由 db.py 迁移逻辑创建（CREATE UNIQUE INDEX ux_watchlist_code），
    # 兼容历史库（create_all 不会为已有表补列/索引）。
    code: str = Field(index=True, max_length=16)          # e.g. "600519.SH"
    name: str = Field(max_length=32)                       # e.g. "贵州茅台"
    price: float = Field(default=0)                        # 现价（元）
    cost: float = Field(default=0)                         # 成本价（元，持仓盈亏计算用）
    change_pct: float = Field(default=0)
    pnl: float = Field(default=0)                           # 今日盈亏（万）
    holding_value: float = Field(default=0)                 # 持仓市值（万）
    position_pct: float = Field(default=0)                  # 仓位占比 %
    sort_order: int = Field(default=0)